Make poll backoff time based.

timsifive opened the pull request:
Requested in https://review.openocd.org/c/openocd/+/6964. I'm making the change here so I have a chance to test it properly before pushing it upstream.

A nice effect of this change is that we avoid some unnecessary polling early on when gdb is connecting and we would poll once every time we send gdb a qXfer packet.

Change-Id: I4bdb9f05839e8c1e01ff6dde49d6589595418095
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Make poll backoff time based.

Requested in https://review.openocd.org/c/openocd/+/6964.
I'm making the change here so I have a chance to test it properly before
pushing it upstream.

A nice effect of this change is that we avoid some unnecessary polling
early on when gdb is connecting and we would poll once every time we
send gdb a qXfer packet.

Change-Id: I4bdb9f05839e8c1e01ff6dde49d6589595418095
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Make polling_interval unsigned.

Hopefully fixes win32 build.

Change-Id: I13d6d475f03bada96b2eb943f2c16df05413d34f
Signed-off-by: Tim Newsome <tim@sifive.com>

JanMatCodasip commented on src/target/target.c:
```diff
@@ -3046,44 +3046,41 @@ static int handle_target(void *priv)
+			/* Increase interval between polling up to 5000ms */
+			target->backoff.interval = MAX(polling_interval,
+					MIN(target->backoff.interval * 2 + 1, 5000));
+			/* Tell GDB to halt the debugger. This allows the user to run
+			* monitor commands to handle the situation. */
```
Cosmetic detail - indent. 

```suggestion
			/* Tell GDB to halt the debugger. This allows the user to run
			 * monitor commands to handle the situation. */
```

JanMatCodasip commented on src/target/target.c:
```diff
@@ -3046,44 +3046,41 @@ static int handle_target(void *priv)
+		/* polling may fail silently until the target has been examined */
+		retval = target_poll(target);
+		if (retval == ERROR_OK) {
+			target->backoff.interval = polling_interval;
```
Just a little favor to whoever will try to understand the code in the future.

```suggestion
			/* Polling succeeded, reset the back-off interval */
			target->backoff.interval = polling_interval;
```

JanMatCodasip commented on src/target/target.c:
```diff
@@ -3046,44 +3046,41 @@ static int handle_target(void *priv)
+		 * allow some slack in the time comparison. Otherwise, if we
+		 * schedule for now+polling_interval, the next poll won't
+		 * actually happen until a polling_interval later. */
+		if (!target->tap->enabled ||
+		    power_dropout ||
+		    srst_asserted ||
+		    timeval_ms() + polling_interval / 2 < target->backoff.next_attempt)
 			continue;
```
An attempt at improving readability:

```suggestion
                const bool time_to_poll = (timeval_ms() + polling_interval / 2) >= target->backoff.next_attempt;
		if (!target->tap->enabled ||
		    power_dropout ||
		    srst_asserted ||
		    !time_to_pool) {
			/* Do not poll this target now */
			continue;
		}
```

timsifive replied:
Done, but slightly differently.

JanMatCodasip reviewed (dismissed):
Checked visually - the changes look all right.

timsifive committed:
Fix comment indent.

Co-authored-by: Jan Matyas <50193733+JanMatCodasip@users.noreply.github.com>
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Add explanatory comment.

Co-authored-by: Jan Matyas <50193733+JanMatCodasip@users.noreply.github.com>
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Make large if() more readable.

Change-Id: Ie43400387ab9f290e744ebaa09786612237e6c7e
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Don't use const on temporary variable.

It's not part of OpenOCD style:
https://review.openocd.org/c/openocd/+/6319/5..10/src/target/aarch64.c#b1500

Change-Id: Ifb612a942507ca5ed8cac3e3ec59e0e14b0298ed
Signed-off-by: Tim Newsome <tim@sifive.com>

GregSavin commented on src/target/target.h:
```diff
@@ -117,8 +117,8 @@ struct gdb_service {
-	int times;
-	int count;
+	int64_t next_attempt;
+	unsigned int interval;
```
With next_attempt and the result from timeval_ms() being signed, and interval field being unsigned, I was wondering if there is the potential for signed/unsigned arithmetic sorts of issues with the calculation "bool poll_needed = timeval_ms() + polling_interval / 2 >= target->backoff.next_attempt;"    But I guess timeval_ms() is never going to return a value with a sign bit of 1, so no problem.

timsifive replied:
Agreed that it's a bit odd, but timeval_ms() returns int64_t and that's why next_attempt is also signed.
